contrib/kubespray/script/pre-check-generator.py

```diff
@@ -73,31 +56,34 @@ def generate_template_file(template_file_path, output_path, map_table):
 
 def main():
     parser = argparse.ArgumentParser()
-    parser.add_argument('-w', '--worker-list-csv', dest="worklist", required=True,
-                        help="worker-list")
-    parser.add_argument('-m', '--master-list-csv', dest="masterlist", required=True,
-                        help="master-list")
-    parser.add_argument('-c', '--configuration', dest="configuration", required=True,
+    parser.add_argument('-l', '--layout', dest="layout", required=True,
+                        help="layout.yaml")
+    parser.add_argument('-c', '--config', dest="config", required=True,
                         help="cluster configuration")
     parser.add_argument('-o', '--output', dest="output", required=True,
                         help="cluster configuration")
     args = parser.parse_args()
 
     output_path = os.path.expanduser(args.output)
 
-    master_list = csv_reader(args.masterlist)
-    head_node = master_list[0]
+    layout = load_yaml_config(args.layout)
+    config = load_yaml_config(args.config)
+
+    masters = list(filter(lambda elem: 'pai-master' in elem and elem["pai-master"] == 'true', layout['machine-list']))
+    workers = list(filter(lambda elem: 'pai-worker' in elem and elem["pai-worker"] == 'true', layout['machine-list']))
+    head_node = masters[0]
```

Comment:
Could we add some check here? e.g. There should be one, and only one master machine; There should be at least 1 worker machine; Worker machine and master machine shouldn't be the same.

Reply:
> Could we add some check here? e.g. There should be one, and only one master machine; There should be at least 1 worker machine; Worker machine and master machine shouldn't be the same.

This is checked in `deployment/clusterObjectModel/mainParser/layout.py`. For now, only the one-master contraint is added, I do not plan to add more since we may remove them very soon.

Reply:
After discussion, the layout should also be checked before k8s installation. We will add this step after the main changed. This will be tracked in #5100